convert2rhel/checks.py

```diff
@@ -16,13 +16,194 @@
+
+    When we scan kernel modules provided by kernel packages
+    it is expensive to check each kernel pkg. Since each new
+    kernel pkg do not deprecate kernel modules we only select
+    the most recent ones.
+
+    :type pkgs: Iterable[str]
+    """
+
+    pkgs_groups = itertools.groupby(
+        pkgs, lambda pkg_name: pkg_name.split(":")[0]
+    )
+    return (
+        max(distinct_kernel_pkgs[1], key=_repos_version_key)
+        for distinct_kernel_pkgs in pkgs_groups
+        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod"))
+    )
+
+
+def get_rhel_supported_kmods(_refresh_system_info=False):
+    """Return set of target RHEL supported kernel modules."""
+    # allowing to rerun resolve_system_info (for the needs of tests)
+    if _refresh_system_info:
+        system_info.resolve_system_info()
+    repoquery_repoids_args = " ".join(
+        "--repoid " + repoid for repoid in system_info.default_rhsm_repoids
+    )
+    # get output of a command to get all packages which are the source
+    # of kmods
+    kmod_pkgs_bytes, _ = run_subprocess(
+        (
+            "repoquery "
+            "--releasever={releasever} "
+            "{repoids_args} "
+            "-f /lib/modules/*.ko*"
+        ).format(
+            releasever=system_info.releasever,
+            repoids_args=repoquery_repoids_args,
+        )
+    )
+    # from these packages we select only the latest one
+    kmod_pkgs = get_most_recent_unique_kernel_pkgs(
+        kmod_pkgs_bytes.decode().rstrip("\n").split()
+    )
+    rhel_kmods = set()
+    for kmod_pkg in kmod_pkgs:
+        # querying obtained packages for files they produces
+        rhel_kmods_bytes, _ = run_subprocess(
+            (
+                "repoquery "
+                "--releasever={releasever} "
+                "{repoids_args} "
+                "-l {pkg}"
+            ).format(
+                releasever=system_info.releasever,
+                repoids_args=repoquery_repoids_args,
+                pkg=kmod_pkg,
+            )
+        )
+        # appending only files which are kmods (ends with ko or ko.xz)
+        rhel_kmods.update(
+            set(
+                _get_kmod_comparison_key(kmod_path)
+                for kmod_path in filter(
+                    lambda path: path.endswith(("ko.xz", "ko")),
+                    rhel_kmods_bytes.decode().rstrip("\n").split(),
+                )
+            )
+        )
+    return rhel_kmods
+
+
+def check_thirdparty_kmods():
+    unsigned_modules, _ = run_subprocess("grep '(' /proc/modules")
+    if unsigned_modules:
+        logger.critical(
+            "Unsigned kernel module(s) has been detected. "
+            "Third-party components are not supported per our "
+            "software support policy\n%s.\n\n "
+            "Uninstall or disable the module(s) and run convert2rhel "
+            "again to continue with the conversion.",
+            LINK_KMODS_RH_POLICY,
+        )
```

Comment:
1. The output of grep is visible in stdout
2. It's not clear which kmod is the problem

```suggestion
    unsigned_modules, _ = run_subprocess("grep '(' /proc/modules")
    module_names = "\n  ".join([mod.split(' ')[0] for mod in unsigned_modules.splitlines()])
    if unsigned_modules:
        logger.critical(
            "Unsigned kernel module(s) has been detected. "
            "Third-party components are not supported per our "
            "software support policy\n%s.\n\n"
            "Uninstall or disable the following module(s) and run convert2rhel "
            "again to continue with the conversion:\n  %s",
            LINK_KMODS_RH_POLICY,
            module_names
        )
```

Reply:
Updated your suggestion @shaded-enmity to remove a dot after the URL. Some resolve the URL with the dot in it

So it will now be

```
CRITICAL - Unsigned kernel module(s) has been detected. Third-party components are not supported per our software support policy
https://access.redhat.com/third-party-software-support
```

instead of

```
CRITICAL - Unsigned kernel module(s) has been detected. Third-party components are not supported per our software support policy
https://access.redhat.com/third-party-software-support.
```

Reply:
applied